Verify credential tutorial (Python) (#3030)
XRPLF/xrpl-dev-portal: docs/tutorials/python/compliance/verify-credential.md

---
seo:
    description: Verify that an account holds a valid credential on the XRP Ledger.
labels:
    - Credentials
---
# Verify Credentials in Python

This tutorial describes how to verify that an account holds a valid [credential](/docs/concepts/decentralized-storage/credentials) on the XRP Ledger, which has different use cases depending on the type of credential and the meaning behind it. A few possible reasons to verify a credential include:

- Confirming that a recipient has passed a background check before sending a payment.
- Checking a person's professional certifications, after verifying their identity with a [DID](/docs/concepts/decentralized-storage/decentralized-identifiers).
- Displaying a player's achievements in a blockchain-connected game.

This tutorial uses sample code in Python using the [xrpl-py library](../index.md).

## Prerequisites

- You must have Python installed and know how to run Python code from the command line. Python 3.8 or later is required for xrpl-py.

> - You should have a basic understanding of the XRP Ledger

Reviewer: ```suggestion
- You should have a basic understanding of the XRP Ledger.
```

- The credential you want to verify should exist in the ledger already, and you should know the addresses of both the issuer and the holder, as well as the official credential type you want to check.
<!-- Uncomment when both are merged:
     - For sample code showing how to create credentials, see [Build a Credential Issuing Service](../build-apps/credential-issuing-service.md).
-->

## Setup

First, download the complete sample code for this tutorial from GitHub:

- {% repo-link path="_code-samples/verify-credential/py/" %}Verify Credential sample code{% /repo-link %}

Then, in the appropriate directory, set up a virtual environment and install dependencies:

```sh
python -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
```

> This should install an appropriate version of the `xrpl-py` library. There are no other dependencies for this tutorial outside of the Python standard library.

Reviewer: ```suggestion
This installs the appropriate version of the `xrpl-py` library. There are no other dependencies for this tutorial outside of the Python standard library.
```

## Overview

> The Verify Credential sample code consists of one file, `verify_credential.py`, containing two main parts:

Reviewer: ```suggestion
The Verify Credential sample code consists of one file, `verify_credential.py`, and contains two main parts:
```

2. A commandline utility that can be used to verify any credential.

## Usage

To test that you have the code installed and working properly, you can run the commandline utility with no arguments to check the existence of a default credential on Devnet, such as:

```sh
python verify_credential.py
```

If all goes well, you should see output such as the following:

```text
Encoded credential_type as hex: 6D795F63726564656E7469616C
Looking up credential...
{'ledger_index': 'validated', 'method': 'ledger_entry', 'api_version': 2, 'credential': {'subject': 'rsYhHbanGpnYe3M6bsaMeJT5jnLTfDEzoA', 'issuer': 'rEzikzbnH6FQJ2cCr4Bqmf6c3jyWLzkonS', 'credential_type': '6D795F63726564656E7469616C'}, 'binary': False}
Found credential:
{'CredentialType': '6D795F63726564656E7469616C', 'Flags': 65536, 'Issuer': 'rEzikzbnH6FQJ2cCr4Bqmf6c3jyWLzkonS', 'IssuerNode': '0', 'LedgerEntryType': 'Credential', 'PreviousTxnID': '7D1257779E2D298C07C7E0C73CD446534B143FBD1F13DB268A878E40FD153B9A', 'PreviousTxnLgrSeq': 803275, 'Subject': 'rsYhHbanGpnYe3M6bsaMeJT5jnLTfDEzoA', 'SubjectNode': '0', 'index': '9603F0E204A8B1C61823625682EB0ECE98A4ECF22FF46CD4845FA9BFA3606B24'}
Credential is valid.
```

> If the code instead reports that the credential was not found when called with no arguments, it's possible that the example credential has been deleted or the Devnet has been reset. If you have another credential you can verify, you can provide the details as commandline arguments. For example:

Reviewer: ```suggestion
If the code reports that the credential was not found when called with no arguments, it's possible that the example credential has been deleted or the Devnet has been reset. If you have another credential you can verify, you can provide the details as commandline arguments. For example:
```

```sh
python verify_credential.py rsYhHbanGpnYe3M6bsaMeJT5jnLTfDEzoA rsYhHbanGpnYe3M6bsaMeJT5jnLTfDEzoA my_credential

> ```

Reviewer: Just a suggestion, feel free to ignore. Might be nice to add the alternative command with an encoded credential as well.

Author: Added to the "Other Examples" tabs below.

A full usage statement is available with the `-h` flag.

### Interactive Shell

You can open an interactive python shell and import the `verify_credential` function, as in the following example:

```py
>>> from verify_credential import verify_credential
>>> from xrpl.clients import JsonRpcClient
>>> client = JsonRpcClient("https://s.devnet.rippletest.net:51234/")
>>> verify_credential(client, issuer="rEzikzbnH6FQJ2cCr4Bqmf6c3jyWLzkonS", subject="rsYhHbanGpnYe3M6bsaMeJT5jnLTfDEzoA", credential_type="my_credential")
True
```

You can import the `verify_credential(...)` function into other scripts and use it the same way.

### Other Examples

> The following examples show other possible scenarios. The data for these examples may or may not still be present in Devnet—for example, anyone can delete an expired credential.

Reviewer: ```suggestion
The following examples show other possible scenarios. The data for these examples may or may not still be present in Devnet. For example, anyone can delete an expired credential.
```

### 2. Main function

The `verify_credential(...)` function performs the main work for this tutorial. The function definition and docstring define the parameters:

{% code-snippet file="/_code-samples/verify-credential/py/verify_credential.py" language="py" from="# Main function" before="# Handle function inputs" /%}

The first thing the function does is verify that the user provided a credential type in either the `credential_type` or `credential_type_hex` parameter. The XRP Ledger APIs require the credential type to be hexadecimal, so it converts the user input if necessary:

{% code-snippet file="/_code-samples/verify-credential/py/verify_credential.py" language="py" from="# Handle function inputs" before="# Perform XRPL lookup" /%}

Next, it calls the [ledger_entry method](/docs/references/http-websocket-apis/public-api-methods/ledger-methods/ledger_entry#get-credential-entry) to look up the requested Credential ledger entry:

{% code-snippet file="/_code-samples/verify-credential/py/verify_credential.py" language="py" from="# Perform XRPL lookup" before="# Confirm that the credential has been accepted" /%}

> If it succeeds in finding the credential, the function continues by checking that the credential has been accepted by its holder. (Anyone can issue any credential to anyone else, so a credential is not considered valid unless the subject of the credential has accepted it.)

Reviewer: ```suggestion
If it succeeds in finding the credential, the function continues by checking that the credential has been accepted by its holder. Since anyone can issue a credential to anyone else, a credential is only considered valid if its subject has accepted it.
```
